File: ui/js/core/responder.js
function Responder(name) {
    this.name = name;
    this.handlers = [];
    this.children = [];
    this.parent = null;
    this.active = false;

    this.registerHandler('startChild', (data) => this.startChild(data.target));
    this.registerHandler('stopChild', (data) => this.stopChild(data.target));
    this.registerHandler('propagate', (data) => this.propagate(data.target, data.command));
}

Responder.prototype.start = function () {
    if (this.active) throw 'Cannot start responder because it is already started';
    this.active = true;

    this.didStart();
}

Responder.prototype.didStart = function () { }

Responder.prototype.stop = function () {
    if (!this.active) throw 'Cannot stop responder because it is already stopped';
    this.active = false;

    this.didStop();
}

Responder.prototype.didStop = function () { }

Responder.prototype.handle = function (command) {
    if (!this.active) throw 'Cannot handle command because of inactive state';

    let handler = this.findHandler(command.name);
    
    if (handler) handler.handler(command.data);
    else throw `Handler is not found for command '${command.name}'`;
}

Responder.prototype.propagate = function (target, command) {
    let child = this.findChild(target);

    if (child) child.handle(command);
    else throw `Target '${target}' is not found for propagation`;
}

Responder.prototype.startChild = function (target) {
    let child = this.findChild(target);

    if (child) child.start();
    else throw `Cannot find and start child '${target}'`;
}

Responder.prototype.stopChild = function (target) {
    let child = this.findChild(target);

    if (child) child.stop();
    else throw `Cannot find and stop child '${target}'`;
}

Responder.prototype.reply = function (command) {
    if (!this.active) throw 'Cannot send reply from inactive responder';
    if (!this.parent) throw 'Cannot send reply because there is no appropriate parent';

    let name = this.name;
    this.parent.reply({
        name: 'propagate',
        data: {
            target: name,
            command: command
        }
    });
}

Responder.prototype.addChild = function (child) {
    this.children.push(child);
    child.parent = this;
}

Responder.prototype.findChild = function (name) {
    for (let child of this.children)
        if (child.name == name) return child;

    return null;
}

Responder.prototype.registerHandler = function (name, handler) {
    this.handlers.push({name: name, handler: handler});
}

Responder.prototype.findHandler = function (name) {
    for (let handler of this.handlers)
        if (handler.name == name) return handler;

    return null;
}

File: hpns/UI/Core/Responder.cs
using System;
using System.Collections.Generic;
using System.Dynamic;

namespace HPNS.UI.Core
{
    public class Responder
    {
        private class CommandHandler
        {
            public string CommandName { get; }

            public Action<dynamic> Handler { get; }

            public CommandHandler(string commandName, Action<dynamic> handler)
            {
                CommandName = commandName;
                Handler = handler;
            }
        }
        
        private List<CommandHandler> _handlers = new List<CommandHandler>();
        private List<Responder> _children = new List<Responder>();

        public string Name { get; }
        public Responder Parent { get; set; }

        public Responder(string name)
        {
            Name = name;
            
            RegisterHandler("propagate", data => Propagate(data.target, data.command));
        }

        public void Handle(dynamic command)
        {
            foreach (var handler in _handlers)
            {
                if (handler.CommandName != command.name) continue;

                handler.Handler(DoesPropertyExist(command, "data") ? command.data : null);
                return;
            }
            
            throw new Exception($"Handler is not found for command '{command.name}'.");
        }

        public void AddChild(Responder child)
        {
            _children.Add(child);
            child.Parent = this;
        }

        protected virtual bool Reply(object command)
        {
            if (Parent == null)
                throw new Exception("Cannot send reply because there is no appropriate parent.");

            return Parent.Reply(new
            {
                name = "propagate",
                data = new {target = Name, command = command}
            });
        }

        protected void RegisterHandler(string commandName, Action<dynamic> handler)
        {
            _handlers.Add(new CommandHandler(commandName, handler));
        }

        private void Propagate(string target, dynamic command)
        {
            foreach (var child in _children)
            {
                if (child.Name != target) continue;

                child.Handle(command);
                return;
            }

            throw new Exception($"Target {target} is not found for propagation.");
        }
        
        private static bool DoesPropertyExist(dynamic obj, string name)
        {
            if (obj is ExpandoObject)
                return ((IDictionary<string, object>) obj).ContainsKey(name);

            return obj.GetType().GetProperty(name) != null;
        }
    }
}